feat: react agent and host multi-agent can be composed directly as AnyGraph

shentongmartin opened the pull request:
#### What type of PR is this?
feat

#### (Optional) Translate the PR title into Chinese.
react agent 和 host multi-agent 支持导出为 AnyGraph，可直接进入编排。

#### description:

refactor agents to support graph composition

- Implement GraphFlow interface for react.Agent and host.MultiAgent
- Add support for proper node path designation in subgraphs
- Maintain backward compatibility for direct usage

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cloudwego/eino?pullRequest=72) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cloudwego/eino?pullRequest=72) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cloudwego/eino?pullRequest=72) it.</sub>

luohq09 commented on flow/agent/multiagent/host/types.go:
```diff
@@ -32,29 +33,84 @@ import (
+	return err
+}
+
+func (ma *MultiAgent) ToLambda(ctx context.Context) (*compose.Lambda, error) {
```
为什么同时提供2种？

luohq09 replied:
ToLambda和ToGraph

shentongmartin replied:
去掉了 ToLambda

luohq09 commented on flow/agent/react/react.go:
```diff
@@ -123,7 +124,11 @@ func firstChunkStreamToolCallChecker(_ context.Context, sr *schema.StreamReader[
+	graph            *compose.Graph[[]*schema.Message, *schema.Message]
+	graphCompileOpts []compose.GraphCompileOption
+	compileOnce      sync.Once
+	compileError     error
```
为什么需要加这两个

shentongmartin replied:
去掉了 compileOnce 和 compileError

luohq09 commented on flow/agent/react/react.go:
```diff
@@ -311,10 +318,28 @@ func (r *Agent) Generate(ctx context.Context, input []*schema.Message, opts ...a
+	return r.graph, r.graphAddNodeOpts
+}
+
+func (r *Agent) ConvertOptions(_ context.Context, path *compose.NodePath, opts ...agent.AgentOption) []compose.Option {
```
感觉不需要提供

luohq09 replied:
因为本来用户提供的就是compose.Option，是转成了AgentOption

shentongmartin replied:
multiagent 定义了自己的 callback：

`type MultiAgentCallback interface { // nolint: byted_s_interface_name
	OnHandOff(ctx context.Context, info *HandOffInfo) context.Context
}`

表示一个 agent 间的跳转事件。有一个对应的 option：WithAgentCallbacks

这个 Callback handler 在实际 graph run 的时候会转成 callbacks.Handler。

这个 ConvertOptions 方法主要是处理这种情况：一个 flow 定义了自己的 ImplSpecificOption，但是在 graph run 的时候是转成 compose.Option 来执行。

shentongmartin replied:
去掉了，提供了一个 ConvertCallbackHandlers 工具函数

luohq09 commented on flow/types.go:
```diff
@@ -0,0 +1,42 @@
+// 1. One less layer of conversion, making it simple and straightforward.
+// 2. The internal graph structure is visible to the entire graph, and compose.Option can be directly designated to the internal graph nodes.
+// 3. The association between the parent and child graphs is exported in the GraphCompileCallback.
+type GraphFlow[TOption any] interface {
```
这个interface的作用是什么

shentongmartin replied:
主要是起到说明的作用：
1. 明确的说明 ExportGraph 的用法
2. 显式的声明 GraphFlow 这种类型的 Flow 可以作为子图编排
3. 作为后面其他类似 flow 实现的一个约束和参考

luohq09 replied:
感觉应该不需要

shentongmartin replied:
去掉了

shentongmartin commented:
> ## ⚠️ Breaking API Changes Detected
> This pull request contains breaking API changes that may affect compatibility:
> 
> ```
> uncompatible api changes found: 
> github.com/cloudwego/eino/flow/agent/multiagent/host
>   Incompatible changes:
>   - MultiAgent: old is comparable, new is not
>   Compatible changes:
>   - (*MultiAgent).ExportGraph: added
>   - (*MultiAgent).HostNodeKey: added
>   - ConvertCallbackHandlers: added
> 
> github.com/cloudwego/eino/flow/agent/react
>   Incompatible changes:
>   - Agent: old is comparable, new is not
>   Compatible changes:
>   - (*Agent).ExportGraph: added
> ```
> 
> ### Required Actions:
> 1. Please review these changes carefully
> 2. Update the user documentation to reflect these API changes
> 3. Add a note in the changelog about these breaking changes

The breaking API detected is for react.Agent and host.MultiAgent to no longer be comparable. Since no one will actually compare these two structs, this will have no impact on existing users. We will proceed to merge and treat the changes in this PR as compatible.
